api/s2n.h

```diff
@@ -596,6 +596,258 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
+ * 
+ * # Safety
+ * 
+ * Use `s2n_offered_psk_free` to free the memory allocated to the s2n offered pre-shared key object created by this API. 
+ *
+ * @return struct s2n_offered_psk* Returns a pointer to the newly created offered pre-shared key object.
+ */
+S2N_API 
+struct s2n_offered_psk* s2n_offered_psk_new();
+
+/**
+ * Frees the memory associated with the `s2n_offered_psk` object.
+ *
+ * @param psk A pointer to the `s2n_offered_psk` object to be freed.
+ */
+S2N_API 
+int s2n_offered_psk_free(struct s2n_offered_psk **psk);
+
+/**
+ * Gets the pre-shared key identity and pre-shared key identity length for a given offered pre-shared key object. 
+ * 
+ * @param psk A pointer to the offered pre-shared key object being read.
+ * @param identity The pre-shared key identity being obtained.
+ * @param size The length of the pre-shared key identity being obtained.
+ */
+S2N_API 
+int s2n_offered_psk_get_identity(struct s2n_offered_psk *psk, uint8_t** identity, uint16_t *size);
+
+struct s2n_offered_psk_list;
+
+/**
+ * Checks whether the offered pre-shared key list has an offered psk object next in line in the list.
+ * An offered pre-shared key list contains all the pre-shared keys offered by the client for the server to select.
+ *
+ * If an offered pre-shared key exists next in line, this API returns True, otherwise False. 
+ * 
+ * @param psk_list A pointer to the offered pre-shared key list being read.
+ * @return bool A boolean value representing whether an offered psk object is present next in line in the offered pre-shared key list.
+ */
+S2N_API 
+bool s2n_offered_psk_list_has_next(struct s2n_offered_psk_list *psk_list);
+
+/**
+ * Obtains the next offered pre-shared key object from the list of offered pre-shared keys. Use `s2n_offered_psk_list_has_next` 
+ * prior to this API call to determine if an offered pre-shared key exists next in the list. 
+ * 
+ * @param psk_list A pointer to the offered pre-shared key list being read.
+ * @param psk A pointer to the next offered pre-shared key object being obtained.
+ */
+S2N_API 
+int s2n_offered_psk_list_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+
+/**
+ * Resets the offered pre-shared key list pointer to the initial pre-shared key object in the order 
+ * in which the pre-shared key objects were originally added to the list by the client.
+ *  
+ * @param psk_list A pointer to the offered pre-shared key list being reread.
+ */
+S2N_API 
+int s2n_offered_psk_list_reread(struct s2n_offered_psk_list *psk_list);
+
+/**
+ * Chooses a pre-shared key from the offered pre-shared key list to be used for the connection.  
+ * This API matches the pre-shared key identity received from the client against the server's known pre-shared key identities 
+ * list, in order to choose the pre-shared key to be used for the connection. If the pre-shared key identity sent from the client 
```

Comment:
I worry this is more implementation details than what the user cares about. They choose a PSK, we use that PSK. They probably don't care that we have to perform matching in order to get the server version of that PSK.

Reply:
I think this detail is important here as this API is not required to be used by default and `They choose a PSK, we use that PSK` only if the callbacks are preferred by the customer - in which case the customers using it would want to know what the API logic is.

Reply:
I'm not sure I understand the reasoning. This API is required if the callback is set, and I'm not sure how that affects the level of detail in the documentation.

Are you trying to communicate to the customer that they need to use append_psk to set up the PSK they choose on the server? If so, I'd be explicit about that. An example in the usage guide will probably also be helpful.

Reply:
The Usage Guide will make this clear:
```
Once the external PSK object is created and updated to include the PSK identity and secret, 
the caller can add the external PSK to the connection  using `s2n_connection_append_psk` API. 
This API sets up the list of PSKs on the connection that will later be used 
by the pre-shared key extension to select a pre-shared key for the connection.
```